### PR TITLE
etcd_worker: batch etcd patch

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve? 
This PR is an improvement to the issue's(#3112 ) fix.

### What is changed and how it works?
1. Batch as many changefeed `patches` as possible into a batch and submit them as an etcd transaction, which can reduce the etcd qps.
~~2. Get the response from etcd `watchCh` as much as possible to reduce the number of etcd worker tick reactor.~~
3. Add metrics for some key points in EtcdWorker.

### Check List 

Tests 

 - Unit test 

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: pkg/orchestrator/etcd_worker.go [hunk 1]
```diff
@@ -283,14 +283,33 @@ func (worker *EtcdWorker) cloneRawState() map[util.EtcdKey][]byte {
 	return ret
 }
 
+var maxBatchPatchSize = 96
+
 func (worker *EtcdWorker) applyPatchGroups(ctx context.Context, patchGroups [][]DataPatch) ([][]DataPatch, error) {
+	getBatchPatches := func(patchGroups [][]DataPatch) ([]DataPatch, int) {
+		batchPatches := patchGroups[0]
```
Comment (ben1009) [inline pkg/orchestrator/etcd_worker.go:290]: check len before get patchGroups[0] ?

File: pkg/orchestrator/etcd_worker.go [hunk 2]
```diff
@@ -283,14 +283,33 @@ func (worker *EtcdWorker) cloneRawState() map[util.EtcdKey][]byte {
 	return ret
 }
 
+var maxBatchPatchSize = 96
+
 func (worker *EtcdWorker) applyPatchGroups(ctx context.Context, patchGroups [][]DataPatch) ([][]DataPatch, error) {
+	getBatchPatches := func(patchGroups [][]DataPatch) ([]DataPatch, int) {
```
Comment (ben1009) [inline pkg/orchestrator/etcd_worker.go:289]: could the func be a named instead ?

File: pkg/orchestrator/etcd_worker.go [hunk 3]
```diff
@@ -283,14 +283,33 @@ func (worker *EtcdWorker) cloneRawState() map[util.EtcdKey][]byte {
...
+			if (len(batchPatches) + len(patches)) > maxBatchPatchSize {
+				break
+			}
+			batchPatches = append(batchPatches, patches...)
+			patchesNum++
+		}
+
+		return batchPatches, patchesNum
+	}
+
```
Comment (ben1009) [inline pkg/orchestrator/etcd_worker.go:305]: any ut cover the change ?

File: pkg/orchestrator/etcd_worker.go [hunk 4]
```diff
@@ -283,14 +283,33 @@ func (worker *EtcdWorker) cloneRawState() map[util.EtcdKey][]byte {
...
 func (worker *EtcdWorker) applyPatchGroups(ctx context.Context, patchGroups [][]DataPatch) ([][]DataPatch, error) {
+	getBatchPatches := func(patchGroups [][]DataPatch) ([]DataPatch, int) {
+		batchPatches := patchGroups[0]
+		patchesNum := 1
+		for _, patches := range patchGroups {
+			if patchesNum == 1 {
+				continue
+			}
+			if (len(batchPatches) + len(patches)) > maxBatchPatchSize {
```
Comment (ben1009) [inline pkg/orchestrator/etcd_worker.go:296]: what if patchGroups[0] already > max, need break down or just leave as it is.

File: pkg/orchestrator/etcd_worker.go [hunk 5]
```diff
@@ -283,14 +283,33 @@ func (worker *EtcdWorker) cloneRawState() map[util.EtcdKey][]byte {
 	return ret
 }
 
+var maxBatchPatchSize = 96
+
 func (worker *EtcdWorker) applyPatchGroups(ctx context.Context, patchGroups [][]DataPatch) ([][]DataPatch, error) {
+	getBatchPatches := func(patchGroups [][]DataPatch) ([]DataPatch, int) {
+		batchPatches := patchGroups[0]
+		patchesNum := 1
+		for _, patches := range patchGroups {
```
Comment (ben1009) [inline pkg/orchestrator/etcd_worker.go:292]: may be use index, the patchesNum could be index + 1 as result, let for loop do the validation

**Review by ben1009 — COMMENTED** [review]
is this pr ready for review, if not ignore  the comments

File: pkg/orchestrator/etcd_worker.go (hunk 4 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:296]: Leave as it is.

File: pkg/orchestrator/etcd_worker.go (hunk 3 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:305]: i will add ut.

File: pkg/orchestrator/etcd_worker.go (hunk 2 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:289]: sure

File: pkg/orchestrator/etcd_worker.go (hunk 1 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:290]: In any condiction we should make sure a `patches` from `patchGroups` does not be splited.

File: pkg/orchestrator/etcd_worker.go (hunk 5 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:292]: good advice~ thanks

File: pkg/orchestrator/etcd_worker.go [hunk 6]
```diff
@@ -128,6 +128,7 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 	rl := rate.NewLimiter(rate.Limit(tickRate), 1)
 	for {
 		var response clientv3.WatchResponse
```
Comment (overvenus) [inline pkg/orchestrator/etcd_worker.go:130]:
```suggestion
```

File: pkg/orchestrator/etcd_worker.go [hunk 7]
```diff
@@ -146,21 +147,46 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 				return errors.Trace(err)
```
Comment (overvenus) [inline pkg/orchestrator/etcd_worker.go:147]:
```suggestion
		case response := <-watchCh:
			// In this select case, we receive new events from Etcd, and call handleEvent if appropriate.
			if err := response.Err(); err != nil {
				return errors.Trace(err)
```

File: pkg/orchestrator/etcd_worker.go (hunk 1 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:316]: the caller will do this check.

File: pkg/orchestrator/batch.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	"go.etcd.io/etcd/clientv3"
+)
+
+const (
+	maxBatchPatchSize    = 64
+	maxBatchResponseSize = 64
+)
+
+// getBatchPatches batch patches
+func getBatchPatches(patchGroups [][]DataPatch) ([]DataPatch, int) {
```
Comment (liuzix) [inline pkg/orchestrator/batch.go:27]: We also need to count bytes here.  https://etcd.io/docs/v3.3/dev-guide/limit/#:~:text=etcd%20is%20designed%20to%20handle,bytes%20flag%20for%20etcd%20server.

File: pkg/orchestrator/batch.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	num := 0
+	totalSize := 0
+	// store changedState of multiple changefeed
+	batchChangedState := make(map[util.EtcdKey][]byte)
+	for i, patches := range patchGroups {
+		changeState, changedSize, err := getChangedState(state, patches)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		// if one changefeed
```
Comment (asddongmen) [inline pkg/orchestrator/batch.go:38]: i will complete it lately.

File: pkg/orchestrator/etcd_worker.go [hunk 10]
```diff
@@ -196,8 +213,11 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 			if !rl.Allow() {
 				continue
 			}
+			startTime := time.Now()
 			// it is safe that a batch of updates has been applied to worker.state before worker.reactor.Tick
 			nextState, err := worker.reactor.Tick(ctx, worker.state)
+			costTime := time.Since(startTime).Seconds()
```
Comment (overvenus) [inline pkg/orchestrator/etcd_worker.go:219]: Could you warn it takes more than 1 second?

File: pkg/orchestrator/metrics.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+package orchestrator
+
+import "github.com/prometheus/client_golang/prometheus"
+
+var (
+	etcdTxnSize = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "ticdc",
+			Subsystem: "etcdWorker",
```
Comment (overvenus) [inline pkg/orchestrator/metrics.go:22]:
```suggestion
			Subsystem: "etcd_worker",
```

File: pkg/orchestrator/metrics.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+			Name:      "etcd_txn_size_bytes",
+			Help:      "Bucketed histogram of of etcd txn size.",
+			Buckets:   prometheus.ExponentialBuckets(1, 2, 18),
+		}, []string{"capture"})
+
+	etcdTxnExecDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "ticdc",
+			Subsystem: "etcdWorker",
```
Comment (overvenus) [inline pkg/orchestrator/metrics.go:31]:
```suggestion
			Subsystem: "etcd_worker",
```

File: pkg/orchestrator/metrics.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+			Name:      "etcd_txn_exec_duration",
+			Help:      "Bucketed histogram of processing time (s) of a etcd txn.",
+			Buckets:   prometheus.ExponentialBuckets(0.002 /* 2 ms */, 2, 18),
+		}, []string{"capture"})
+
+	etcdWorkerTickDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "ticdc",
+			Subsystem: "etcdWorker",
```
Comment (overvenus) [inline pkg/orchestrator/metrics.go:40]:
```suggestion
			Subsystem: "etcd_worker",
```

File: pkg/orchestrator/etcd_worker.go [hunk 14]
```diff
@@ -344,7 +357,11 @@ func (worker *EtcdWorker) applyPatches(ctx context.Context, patches []DataPatch)
 		panic("unreachable")
 	}
 
+	worker.metrics.metricEtcdTxnSize.Observe(float64(size))
+	startTime := time.Now()
 	resp, err := worker.client.Txn(ctx).If(cmps...).Then(ops...).Commit()
+	costTime := time.Since(startTime).Seconds()
```
Comment (overvenus) [inline pkg/orchestrator/etcd_worker.go:363]: Could you warn it takes more than 1 second?

File: pkg/orchestrator/batch.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/pingcap/ticdc/pkg/orchestrator/util"
+)
+
+const (
+	// 1.25 MiB
+	// Ref: https://etcd.io/docs/v3.3/dev-guide/limit/
+	etcdTxnMaxSize = 1024 * (1024 + 256)
+)
+
+func getBatchChangedState(state map[util.EtcdKey][]byte, patchGroups [][]DataPatch) (map[util.EtcdKey][]byte, int, int, error) {
```
Comment (liuzix) [inline pkg/orchestrator/batch.go:28]: Can you add comments on the meaning of the return values?

**Review by liuzix — APPROVED** [review]

**Review by overvenus — APPROVED** [review]

File: pkg/orchestrator/batch.go (hunk 15 above)
Comment (asddongmen) [reply pkg/orchestrator/batch.go:33]: done

**Pull request merged**